libraries/config.rb

```diff
@@ -0,0 +1,150 @@
+      include Grafana::Cookbook::Utils
+
+      GLOBAL_CONFIG_PROPERTIES_SKIP = %i(conf_directory config_file cookbook source source_ldap source_env owner group filemode sensitive extra_options).freeze
+      private_constant :GLOBAL_CONFIG_PROPERTIES_SKIP
+
+      private
+
+      def resource_properties
+        properties = instance_variable_defined?(:@new_resource) ? new_resource.class.properties(false).keys : self.class.properties(false).keys
+        Chef::Log.debug("resource_properties: Got properties from resource: #{properties.join(', ')}")
+        properties.reject! { |p| GLOBAL_CONFIG_PROPERTIES_SKIP.include?(p) }
+
+        if respond_to?(:resource_config_properties_skip)
+          Chef::Log.debug("resource_properties: Resourced defined skip properties: #{resource_config_properties_skip.join(', ')}")
+          properties.reject! { |p| resource_config_properties_skip.include?(p) }
+        end
+
+        Chef::Log.info("resource_properties: Filtered properties: #{properties.join(', ')}")
+        properties
+      end
+
+      def accumulator_config(action, key, value)
+        path = if respond_to?(:resource_config_path_override)
+                 raise ArgumentError, 'Path override should be specified as an Array' unless resource_config_path_override.is_a?(Array)
+                 resource_config_path_override
+               else
+                 resource_default_config_path
+               end
+
+        config_hash = accumulator_config_path_init(*path)
+        Chef::Log.debug("Append to config key #{key}, value #{value} on path #{path.map { |p| "['#{p}']" }.join}")
+
+        case action
+        when :set
+          config_hash[key] = value
+        when :append
+          config_hash[key] ||= ''
+          config_hash[key].concat(value.to_s)
+        when :push
+          config_hash[key] ||= []
+          config_hash[key].push(value)
+        when :delete
+          config_hash.delete(key) if config_hash.key?(key)
+        else
+          raise ArgumentError, "Unsupported accumulator config action #{action}"
+        end
+      end
+
+      def config_template_exist?
+        Chef::Log.debug("config_template_exist?: Checking for config file #{new_resource.config_file}")
+        config_resource = !find_resource!(:template, ::File.join(new_resource.config_file)).nil?
+
+        Chef::Log.debug("config_template_exist?: #{config_resource}")
+        config_resource
+      rescue Chef::Exceptions::ResourceNotFound
+        Chef::Log.info("config_template_exist?: Config file #{new_resource.config_file} ResourceNotFound")
+        false
+      end
+
+      def init_config_template(load_existing = false)
+        return false if config_template_exist?
+
+        Chef::Log.debug("init_config_template: Creating config template resource for #{new_resource.config_file}")
+
+        with_run_context(:root) do
+          declare_resource(:chef_gem, 'deepsort') { compile_time true } unless gem_installed?('deepsort')
```

Comment:
Do we still need compile time now we're using unified_mode?

Reply:
this is going into the root run_context which doesn't have unified_mode

declaring compile_time resources in the root run context is really weird though, particularly to install gems.  the idempotency check here is also a codesmell.

i don't know why this isn't just `chef_gem 'deepsort'`

if it needs to be done at the outer context with idempotency around the resource and forced to compile time:

`find_resource(:chef_gem, 'deepsort') { compile_time true }`